Faster and more robust editor asset management

kean opened the pull request:
## What?

The subsequent editor loads are now **~1.3s** faster. The first launch is **~1s** faster.

There are a few changes compared to the previous implementation:

- Move code that fetches assets (manifest and files) to the new `EditorService`. The assets manifest is now fetched in parallel with the editor settings.
- Only `EditorService` now triggers the actual download of the assets. The assets are put into a `GutenbergKit/{siteURL}/assets` folder.
- Remove `EditorAssetsProvider` and instead inject the plugin styles and scripts in `window.GBKit` when launching the editor (the same way it's done for editor settings). It happens only if `EditorService` verifies that all the required files exist on disk. If not, the plugins are disabled until the next time.
- `EditorService` reads editor settings and assets manifest in one atomic operation. When new manifest is loaded, it replaces the current one only after fetching all assets for it. Until then, the editor will always use the previous version of the manifest.
- Remove `editorSettings` from `EditorConfiguration`. Add them to internal `EditorDependencies` types fetched by `EditorViewController`

**Minor changes:**
- Remove `editorAssetsEndpoint` parameter from `EditorConfiguration`
- Make logging a cross-cutting concern, so it could be configured once and would be easier to use across GBK 

**Planned future changes:**

- Cleanup unused assets on launch to make sure the disk usage does not grow
- Improve how and when refresh and assets download is triggered
- Cover with unit tests and generate the same code in Kotlin


## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a post or page. -->
<!-- 2. Insert a heading block. -->
<!-- 3. etc. -->

### Accessibility Testing Instructions
<!-- How can you test the changes by using a keyboard/screen reader only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->

## Screenshots or screencast <!-- if applicable -->

Before / After 

**Note**: the *preparing editor* message disappears quicker because "fetch editor settings" is now a step inside `EditorViewController`.


https://github.com/user-attachments/assets/bdf6cbbd-4561-464a-bfc1-61a7ef02901b

kean committed:
Add EditorService

kean committed:
Remove editorAssetsEndpoint parameter from EditorConfiguration

kean committed:
Move assets management to EditorService

kean committed:
Add logging

kean committed:
Add a way to clear loaded assets

kean committed:
Inject manifest directly without EditorAssetsProvider

kean committed:
Simplify CachedAssetSchemeHandler

kean committed:
Move private methods below

kean committed:
Add manifest verification

kean committed:
Make refresh non-throwing

kean committed:
Add State file as an indicator where the request finished or not

kean committed:
Fix build errors

kean committed:
Remove EditorAssetsManifest

kean committed:
Fix build errors

kean committed:
Add shared EditorService instances

kean committed:
Cleanup

kean committed:
Do not refresh too often

kean commented on ios/Sources/GutenbergKit/Sources/Service/EditorAssetsManifest.swift:
```diff
@@ -0,0 +1,106 @@
+import CryptoKit
+import SwiftSoup
+
+struct EditorAssetsManifest: Codable {
```
This is an existing `struct`. I made no changes to it – just renamed the file.

kean committed:
Use Application directory instead of Documents

kean committed:
Make logging a cross-cutting concern

kean committed:
Cleanup

kean committed:
Replace safeFilename with sha1

kean committed:
Cleanup

kean committed:
Trigger refresh automatically in the background after teh editor is loaded

dcalhoun commented on ios/Demo-iOS/Sources/Views/AppRootView.swift:
```diff
@@ -109,13 +99,15 @@ struct AppRootView: View {
+        let config = EditorConfigurationBuilder()
             .setShouldUsePlugins(false)
-            .setSiteUrl("")
+            .setSiteUrl("bundled")
```
What is the purpose of this change? Is this merely an identifier of why the configuration is empty?

kean replied:
Oh, I added it as part of another changes I later reverted. I'll revert this as well.

kean replied:
Reverted

dcalhoun commented on ios/Sources/GutenbergKit/Sources/EditorConfiguration.swift:
```diff
@@ -26,16 +26,10 @@ public struct EditorConfiguration: Sendable {
     public let locale: String
     /// Enables the native inserter UI in the editor
     public let isNativeInserterEnabled: Bool
-    /// Endpoint for loading editor assets, used when enabling `shouldUsePlugins`
-    public var editorAssetsEndpoint: URL?
```
Why are we removing `editorAssetsEndpoint`? I believe it's useful in the future whenever we change the REST API endpoint used, it also allows alternatives for third-parties as needed.

kean replied:
I suggest to add it again in the future if it's needed. For now, I kept it simple as `EditorService` knows what endpoints to use, and the code in the app was duplicating the same endpoint.

dcalhoun replied:
Actually, I believe we must retain this to support WPCOM sites. They require setting a namespace endpoint—e.g., `https://public-api.wordpress.com/wpcom/v2/sites/<site_id>/editor-assets`. 

From my testing, plugin loading currently fails for WPCOM Simple sites.

kean replied:
Good catch. So far, I've only tested it with an Atomic site as the demo doesn't seem to support Simple sites:

<img width="320" alt="Screenshot 2025-12-02 at 9 46 59 AM" src="https://github.com/user-attachments/assets/826620ef-e2f9-4518-bb95-57fc07c8d21a" />

I'll probably revert it for now, but there has to be a better way to do it. I'm not completely happy with how the configuration is spread out between GBK and the app. I'm leaning towards centralizing this knowledge in GBK so it would work in both the app and the demo. The [configuration code](https://github.com/wordpress-mobile/WordPress-iOS/blob/4bc758160f4f26bdc9164caad3e262faf9d59c23/WordPress/Classes/Utility/Editor/EditorConfiguration%2BBlog.swift#L10) is also a bit hard to follow at the moment, but it seems like it's better be left for future PRs.

dcalhoun replied:
Correct, the demo app does not support WPCOM sites. I tested via integrating with the WordPress-iOS app. 

I open to refactoring the API for configuring and starting the editor. 

As much as possible, we should avoid WPCOM/Jetpack references in the GutenbergKit code base. There are few that exist today, but they need to be removed in the future.

dcalhoun commented on ios/Sources/GutenbergKit/Sources/EditorConfiguration.swift:
```diff
@@ -26,16 +26,10 @@ public struct EditorConfiguration: Sendable {
     public let isNativeInserterEnabled: Bool
-    /// Endpoint for loading editor assets, used when enabling `shouldUsePlugins`
-    public var editorAssetsEndpoint: URL?
-    /// Logs emitted at or above this level will be printed to the debug console
-    public let logLevel: LogLevel
```
Are host apps still able to configure the log levels emitted from the JavaScript logger utility? I believe that to be beneficial.

kean replied:
Yes, I still inject the same log level in JS: `logLevel: '\(EditorLogger.logLevel.rawValue)',`

dcalhoun commented on ios/Sources/GutenbergKit/Sources/EditorLogging.swift:
```diff
@@ -0,0 +1,45 @@
+}
+
+func log(_ level: EditorLogLevel, _ message: @autoclosure () -> String) {
+    guard level.priority >= EditorLogger.logLevel.priority,
```
The [previous idea](https://github.com/wordpress-mobile/GutenbergKit/pull/202#discussion_r2460868311) was allowing the JavaScript logger to control this to avoid duplication. I suppose we must add this now so that we can also log directly from the iOS code, correct?

kean replied:
>  also log directly from the iOS code, correct?

Yes, it's for logging directly from Swift. It look like there is no way around duplicating it. I think it's fine.

dcalhoun commented on ios/Sources/GutenbergKit/Sources/EditorViewController.swift:
```diff
@@ -157,14 +150,15 @@ public final class EditorViewController: UIViewController, GutenbergEditorContro
             },
-            logLevel: '\(configuration.logLevel)'
+            logLevel: '\(EditorLogger.logLevel.rawValue)',
+            manifest: \(dependencies?.manifest ?? "undefined")
```
Any performance implications from adding this? 🤔  I suppose it is no larger than `editorSettings`, eh? 

I wonder if it's worth spreading the properties over the top-level object. I could see one day allowing GBK configuration to control `allowedBlockTypes`.

kean replied:
> Any performance implications from adding this? 🤔 I suppose it is no larger than editorSettings, eh?

I wouldn't expect it to be slower than the current implementation, but I already had some second thoughts yesterday. Is it safe to pass these larger strings directly in JS this way? Does it need escaping? Is there a more robust way to do it? Should we pass the whole pre-processed manifest as a separate object? Both settings and manifest seem to have `<scripts> ` and `<styles>`, is there a better way to inject them? What about `allowedBlockTypes`? Should the app combine them (for core and plugins)?

dcalhoun replied:
I agree, my original performance concerns are likely invalid; I figured I'd share my thought nonetheless. 

I do not have a better idea for injecting them at this time. 

Regarding escaping, maybe there is benefit. That said, if the itself relies upon the same scripts we are retrieving, would site already be compromised? Is the mobile editor exposing anything that site does not already? 

For `allowedBlockTypes`, I envision this eventually becoming an editor configuration option rather than a value returned from the editor assets endpoint. I consider its current location a stopgap solution while various blocks do not function in the mobile editor.

dcalhoun commented on ios/Sources/GutenbergKit/Sources/EditorViewController.swift:
```diff
@@ -247,8 +246,16 @@ public final class EditorViewController: UIViewController, GutenbergEditorContro
+            do {
+                self.dependencies = try await service.dependencies(for: configuration)
+            } catch {
+                print("Failed to setup editor environment, continuing with the default or cached configuration:", error)
```
```suggestion
                print("Failed to set up the site's editor configuration, continuing with the default or cached configuration:", error)
```

dcalhoun commented on ios/Sources/GutenbergKit/Sources/EditorViewController.swift:
```diff
@@ -483,14 +490,19 @@ public final class EditorViewController: UIViewController, GutenbergEditorContro
+        Task {
+            await EditorService.shared(for: configuration.siteURL)
+                .refresh(configuration: configuration)
+        }
```
Is this not invoked from opening the view a few lines above?

kean replied:
Not without a delay, but I refactored it now, so these lines are no longer needed.

dcalhoun commented on src/utils/bridge.js:
```diff
@@ -336,13 +337,15 @@ export function awaitGBKitGlobal( timeoutMs = 3000 ) {
 
-	const { siteApiRoot, editorAssetsEndpoint, authHeader } = getGBKit();
+	// Android fallback - fetch from API
+	const { siteApiRoot, editorAssetsEndpoint, authHeader } = gbkit;
```
I don't believe there is a downside to de-structuring all values up front. 

```suggestion
	const { manifest, siteApiRoot, editorAssetsEndpoint, authHeader } = getGBKit();

	// iOS provides manifest directly in GBKit configuration
	if ( window.webkit && manifest ) {
		return manifest;
}

	// Android fallback - fetch from API
```

Alternatively, I don't believe there is a downside to invoking `getGBKit()` multiple times, as it [early returns](https://github.com/wordpress-mobile/GutenbergKit/blob/dac7dc6beda3915fbca83dc09bfc1aee721a440d/src/utils/bridge.js#L195-L197) with the cached value on the `window`. 


```suggestion
	const { manifest } = getGBKit();

	// iOS provides manifest directly in GBKit configuration
	if ( window.webkit && gbkit.manifest ) {
		return gbkit.manifest;
}

	// Android fallback - fetch from API
	const { siteApiRoot, editorAssetsEndpoint, authHeader } = getGBKit();
```

kean replied:
I reverted this change together with how manifest is passed to the editor (now it again uses separate post message scheme).

dcalhoun replied:
Would you please explain the rationale for reverting for my own education? 🙇🏻‍♂️

dcalhoun commented on ios/Sources/GutenbergKit/Sources/Service/EditorService.swift:
```diff
@@ -1,78 +1,171 @@
 
     /// Refresh the editor resources.
-    public func refresh() async throws {
+    /// Will not refresh more often than once every 30 seconds.
```
Is the specific context where rapid invocations might occur or is merely a safeguard?

kean replied:
When you open the app, the editor (`EditorService`) now refreshes all these resources as part of `warmup`. This code makes sure that if you open the editor shortly after opening the app, we don't refresh it twice in a row.

dcalhoun commented on ios/Sources/GutenbergKit/Sources/EditorLogging.swift:
```diff
@@ -0,0 +1,45 @@
+          let logger = EditorLogger.shared else {
+        return
+    }
+    logger.log(level, message())
```
The JavaScript loggers [prepends](https://github.com/wordpress-mobile/GutenbergKit/blob/dac7dc6beda3915fbca83dc09bfc1aee721a440d/src/utils/logger.js#L75) `[GBK]` as a consistent identifier to all messages. Does it make sense to do so here or is that counter to iOS/Xcode practices?

kean replied:
I'll add it in the logger implementation in the host app. For `OSLog` in the demo, there is no need for a prefix as it has a "subsystem" name that you can display next to the individual log messages and filter messages by.
<img width="1204" height="272" alt="Screenshot 2025-11-21 at 10 08 54 AM" src="https://github.com/user-attachments/assets/6a31b61c-224c-41ba-a2d2-3a66f46c2434" />

dcalhoun reviewed (commented):
Thank you for creating this! 

This makes sense to me a high level. It feels like a worthwhile improvement overall, especially given the performance improvements. 

I left a few thoughts to consider. Many are for bettering my own understanding.

kean committed:
Fix typos

kean committed:
Add performance tracking for dependencies

kean committed:
Simlify warmup for editor

kean committed:
Make dependencies func non-throwing

kean committed:
Fix typos in EditorManifestTests

kean committed:
Add missing startEditorSetup call

kean committed:
Swift to Application Support directory

kean committed:
Add EditorSeviceTests

kean committed:
Add loadsSettingsWhenAssetFails

kean committed:
Simplify fetchAssets

kean committed:
Add upgradesManifestOnVersionChange

kean committed:
Refactor tests

kean committed:
Simplify getCachedAsset

kean committed:
Add assertion to check that the assets are actually available

kean committed:
Cache processed manifest (takes ~100ms)

kean committed:
Add cleanupOrphanedAssets

kean committed:
Fix concurrency warnings

kean commented:
I made a few more changes:

- Added unit tests and fixed a couple of issues (it explains most of the line additions)
- Updated EditorService to store processed manifest on disk (it takes around 100ms to generate it)
- Added support for cleaning up orphaned asset files, so we don’t take too much space on disk

With these changes, this whole step now takes 2ms.

I had to switch to a different task today and didn't have time to review the final PR. I'll do it when I have time. It should be good.

kean committed:
Revert createBundledConfiguration change

kean committed:
Restore EditorAssetsProvider and revert manifest injection changes

This restores the EditorAssetsProvider to handle manifest loading via
WebKit message handler instead of injecting it directly into the
configuration.

Changes:
- Restored EditorAssetsProvider.swift to handle manifest requests
- Added EditorAssetsLibrary wrapper around EditorService
- Removed manifest property from EditorConfiguration and EditorDependencies
- Updated EditorViewController to use EditorAssetsProvider
- Updated bridge.js to use message handler for iOS manifest fetching
- Removed manifest from GBKit configuration

kean committed:
Fix LogLevel references to use EditorLogLevel

kean committed:
Update tests

kean committed:
Add more unit tests

kean commented:
I updated the unit tests and added a few more. It should be good for review and merge now.

dcalhoun commented on ios/Demo-iOS/Sources/Views/AppRootView.swift:
```diff
@@ -40,9 +40,13 @@ struct AppRootView: View {
-            case .none: self.activeEditorConfiguration = nil
+            case .bundledEditor:
+                let config = createBundledConfiguration()
+                activeEditorConfiguration = config
```
Likely worthwhile consistently using `self` in this `switch`.
```suggestion
                self.activeEditorConfiguration = config
```

dcalhoun commented on ios/Demo-iOS/Sources/Views/EditorView.swift:
```diff
@@ -202,6 +198,9 @@ private final class EditorViewModel {
     NavigationStack {
-        EditorView(configuration: .default)
+        let config = EditorConfigurationBuilder()
+            .setSiteUrl("preview")
```
Should we revert this given the `"bundled"` change was reverted?

kean replied:
Reverted

dcalhoun commented on ios/Sources/GutenbergKit/Sources/Service/EditorService.swift:
```diff
@@ -81,34 +198,289 @@ public actor EditorService {
+        let localURL = assetsDirectoryURL.appendingPathComponent(cachedFilename(for: urlString))
+
+        if FileManager.default.fileExists(atPath: localURL.path) {
+            log(.debug, "Found cached asset for: \(url.lastPathComponent) (\(localURL.fileSize.formatted))")
```
This seems to always log `0 KB` for each file in my testing. Is that expected? 

This also occurs with `Cleaned up 2 orphaned assets (Zero KB)`. 

<img width="1116" height="570" alt="Image" src="https://github.com/user-attachments/assets/cf7793ad-2c74-4769-978c-f8a3dff04c1c" />

kean replied:
Fixed (removed these for now)

dcalhoun commented on ios/Demo-iOS/Sources/Services/.swift:
This file has an empty file name. Is that intentional?

kean replied:
Removed

dcalhoun reviewed (commented):
Overall, this tested well for me in the demo app. 

As stated in a new inline comment, I believe we must retain the `editorAssetsEndpoint` configuration. I also noted a few oddities likely worth addressing. 

This may be understood, but we need to open a sibling WordPress-iOS PR integrating these changes.

kean committed:
Remove empty file

kean committed:
Revert the preview change

kean committed:
Restore editorAssetsEndpoint parameter

kean committed:
Remove log for local files

jkmassel reviewed (approved):
After some discussion with @dcalhoun and @kean, I'm approving this to merge and we can deal with breakage later. This and #232 represent some pretty serious refactoring, so we should get them landed then incrementally adjust after.

jkmassel committed:
Clean up the merge mess
